### ISSUE TITLE
Two exact same services one gives 404 :( with Ingress Gateway

### ISSUE THREAD

**Issue by nitinkapur** [istio/istio#24499]
**Bug description**

I currently am running on Kubernetes 1.15 with Istio 1.6 installed. Created a Gateway in Istio-system namespace.

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: ingress-gateway-configuration
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway # use Istio default gateway implementation
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
      - "*"
```

created a deployment exactly copying the  bookinfo example. So created two services and deployments of product page and zzz(my microservice) both in namespace dev.

zzz is a microservice that gives back a json object. So the curl zzz.dev.svc.cluster.local gives back the version number, like below

{
    "version": "1.0.0"
}

```
apiVersion: v1
kind: Service
metadata:
  name: productpage
  labels:
    app: productpage
    service: productpage
spec:
  ports:
  - port: 9080
    name: http
  selector:
    app: productpage
```
---
```
    apiVersion: v1
    kind: Service
    metadata:
      name: zzz
      labels:
        app: zzz
        service: zzz
    spec:
      ports:
      - port: 80
        name: http
      selector:
        app: zzz
```
---
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: productpage-v1
  labels:
    app: productpage
    version: v1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: productpage
      version: v1
  template:
    metadata:
      labels:
        app: productpage
        version: v1
    spec:
      #serviceAccountName: bookinfo-productpage
      containers:
      - name: productpage
        image: docker.io/istio/examples-bookinfo-productpage-v1:1.15.0
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 9080
        volumeMounts:
        - name: tmp
          mountPath: /tmp
      volumes:
      - name: tmp
        emptyDir: {}
```
---
```
apiVersion: apps/v1
kind: Deployment
metadata:
    name: zzz-v1
    labels:
      app: zzz
      version: v1
spec:
    replicas: 1
    selector:
      matchLabels:
        app: zzz
        version: v1
    template:
      metadata:
          labels:
            app: zzz
            version: v1
      spec:
        containers:
        - name: zzz
          image: xxx
          imagePullPolicy: IfNotPresent
          ports:
          - containerPort: 80
          volumeMounts:
          - name: config-volume
            mountPath: /app/config
            readOnly: true
        volumes:
            - name: config-volume
              configMap:
                  name: appsettings-zzz-config
---
```

Now I have 2 services zzz and productpage in namespace dev. Below I create a virtual service for both of them in namesapce dev

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: bookinfo
spec:
  hosts:
  - "*"
  gateways:
  - istio-system/ingress-gateway-configuration
  http:
  - match:
    - uri:
        exact: /productpage
    - uri:
        prefix: /static
    - uri:
        exact: /login
    - uri:
        exact: /logout
    - uri:
        prefix: /api/v1/products
    route:
    - destination:
        host: productpage
        port:
          number: 9080
```

and another one

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: zzz
spec:
  hosts:
  - "*"
  gateways:
  - istio-system/ingress-gateway-configuration
  http:
  - name: "zzz-route"
    match:
    - uri:
        prefix: /zzz
    route:
    - destination:
        host: zzz.dev.svc.cluster.local
        port:
          number: 80
    timeout: 3.0s
```

After this if I try to access productpage from Ingress Gateway it works

htttp://Ingress-Gateway-Address/productpage

![image](https://user-images.githubusercontent.com/10778343/83986881-4f199600-a90c-11ea-9034-e2586df6a54c.png)

But is I do the same for my another service zzz it gives me a 404 error

htttp://Ingress-Gateway-Address/zzz

Using the below command I see that the service does gives back it's version number 
kubectl exec -it ZZZ_POD -n dev -c istio-proxy /bin/bash 

and 

kubectl exec productpage-v1-7f4f958b8c-2tr5d -n dev -c istio-proxy -- curl -v 10.133.106.101:80

does gives the below output, so looks it is available from outside.

![image](https://user-images.githubusercontent.com/10778343/84044112-224d9900-a975-11ea-83a4-828e244cc4ed.png)

When I hit the Gateway directly below is the output. I have hidden the actual service name(please assume it is zzz ) and load balancer address.

![image](https://user-images.githubusercontent.com/10778343/84043809-bec36b80-a974-11ea-96f8-529124271879.png)

Since both the services are setup exactly the same way and one service is returning webpage but the other is returning 404, instead it should return a json object with a version, it is very strange. 

Can you guys please help me triage? I am new to Istio.

**Expected behavior**

Should get the version back

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

**How was Istio installed?**

I installed using with a custom configuration see attached. Basically used Istioctl and generated manifests and then tweaked them to install below features

![image](https://user-images.githubusercontent.com/10778343/83987014-d0712880-a90c-11ea-8729-007a69a1e84f.png)

**Environment where bug was observed (cloud vendor, OS, etc)**

It is a linux based EKS cluster

[istio-configuration.yaml.zip](https://github.com/istio/istio/files/4743549/istio-configuration.yaml.zip)

**Comment by esnible** [comment]
Try `istioctl analyze` and see if it reports any problems.

You have the prefix `/zzz`.  Your _curl_ is `curl -v :80` which lacks a `/zzz`.

**Comment by nitinkapur** [comment]
Sorry, That was a type, I have replace it with the actual ip address that I tried. 

No issues found with Istioctl

![image](https://user-images.githubusercontent.com/10778343/84040893-3099b600-a971-11ea-9f6f-9e2927d21cf8.png)

**Comment by nitinkapur** [comment]
Strange thing is that if I change the virtual service for zzz to point to productpage

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: zzz
spec:
  hosts:
  - "*"
  gateways:
  - istio-system/ingress-gateway-configuration
  http:
  - match:
    - uri:
        exact: /productpage
    - uri:
        prefix: /static
    - uri:
        exact: /login
    - uri:
        exact: /logout
    - uri:
        prefix: /api/v1/products
    route:
    - destination:
        host: productpage.dev.svc.cluster.local
        port:
          number: 9080
    timeout: 3.0s
```
I start getting a 404!!! lol

![image](https://user-images.githubusercontent.com/10778343/84208691-4ba81b00-aa82-11ea-8dbb-9556938fd598.png)


Why does the above route does not work. But the below route still works!!!

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: bookinfo
spec:
  hosts:
  - "*"
  gateways:
  - istio-system/ingress-gateway-configuration
  http:
  - match:
    - uri:
        exact: /productpage
    - uri:
        prefix: /static
    - uri:
        exact: /login
    - uri:
        exact: /logout
    - uri:
        prefix: /api/v1/products
    route:
    - destination:
        host: productpage
        port:
          number: 9080
```

This is so confusing!!!

**Comment by nitinkapur** [comment]
Yes! you are right. I figured it out now. 

My route was all wrong, I thought the mappings work the same as in Ambassador Edgestack as both are built on Envoy. In Ambassador EdgeStack I can just give a route "/zzz/" and the service name and it point to that service's base endpoint. So http://<GateWayAddress>/zzz/ wil give the version number. Service doesn't has to have /zzz endpoint. 
But looks like here it is not the same with Istio so http://<GateWayAdress>/zzz only works if service has an endpoint named '/zzz'

That was a mistake on my part but thanks to you, I was able resolve it.

The strategy that you have shared above is really good and would help me triage issues in future.

Thanks!
